rmb-sdk-go/peer/router.go

```diff
@@ -0,0 +1,133 @@
+package peer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/tfgrid-sdk-go/rmb-sdk-go"
+	"github.com/threefoldtech/tfgrid-sdk-go/rmb-sdk-go/peer/types"
+)
+
+var (
+	// ErrFunctionNotFound is an err returned if the handler function is not found
+	ErrFunctionNotFound = fmt.Errorf("function is not found")
+)
+
+// Handler is a handler function type
+type HandlerFunc func(ctx context.Context, payload []byte) (interface{}, error)
+
+// Middleware is middleware function type
+type Middleware func(ctx context.Context, payload []byte) (context.Context, error)
+
+type Router struct {
+	handlers map[string]HandlerFunc
+	routes   map[string]Router
+	mw       []Middleware
+}
+
+func NewRouter() Router {
+	return Router{
+		handlers: make(map[string]HandlerFunc),
+		routes:   make(map[string]Router),
+	}
+}
+
+// SubRoute add a route prefix to include more sub routes with handler from it
+func (r *Router) SubRoute(prefix string) Router {
+	if strings.Contains(prefix, ".") {
+		panic("invalid subrouter prefix should not have '.'")
+	}
+
+	sub, ok := r.routes[prefix]
+	if ok {
+		return sub
+	}
+
+	router := NewRouter()
+	r.routes[prefix] = router
+	return router
+}
+
+// WithHandler adds a handler function to a router sub command
+func (r *Router) WithHandler(subCommand string, handler HandlerFunc) {
+	if _, ok := r.handlers[subCommand]; ok {
+		panic("handler function is already registered")
+	}
+
+	r.handlers[subCommand] = handler
+}
+
+// Use adds a middleware to the router
+func (r *Router) Use(mw Middleware) {
+	r.mw = append(r.mw, mw)
+}
+
+func (r *Router) Serve(ctx context.Context, peer Peer, env *types.Envelope, err error) {
+	// parse and call request
+	errReq := env.GetError()
+	if errReq != nil {
+		log.Error().Err(err)
+		return
+	}
+
+	req := env.GetRequest()
+	if req == nil {
+		log.Error().Msg("received a non request envelope")
+		return
+	}
+
+	if env.Schema == nil || *env.Schema != rmb.DefaultSchema {
+		log.Error().Msgf("invalid schema received expected '%s'", rmb.DefaultSchema)
+		return
+	}
+
+	payload := env.Payload.(*types.Envelope_Plain).Plain
+	cmd := env.GetRequest().Command
+
+	response, err := r.call(ctx, cmd, payload)
+
+	// send response
+	if err := peer.SendResponse(ctx, uuid.NewString(), env.Source.Twin, env.Source.Connection, err, response); err != nil {
```

Comment:
response HAS to use the same ID for request, not a new one. Obviously you never tested if your code works. Using a new ID means the caller won't be able to map the response to the request he sent.